Adds a System 2.0 version of Bot Visualizer

liangfok opened the pull request:
This is just a direct port of the System 1.0 Bot Visualizer.

I first verified that the models appear correctly in Drake Visualizer. I then programmed in the expected LCM messages and added logic to ensure the LCM messages received match the expected LCM messages.

## Box
![box-in-drake-visualizer](https://cloud.githubusercontent.com/assets/1388098/17917454/bc6125b8-698a-11e6-89ae-f13470d8a95e.png)

## Capsule
![capsule](https://cloud.githubusercontent.com/assets/1388098/17917457/c1a6cfa0-698a-11e6-934a-9ef056c96034.png)

## Cylinder
![cylinder](https://cloud.githubusercontent.com/assets/1388098/17917456/bf4a59ac-698a-11e6-87d7-256287a614ac.png)

## Sphere

![sphere](https://cloud.githubusercontent.com/assets/1388098/17917505/32fc4bf8-698b-11e6-812c-710b38c71ac8.png)


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/3228)
<!-- Reviewable:end -->

liangfok commented:
The previous CI test passed every test except for Win32. I believe I've fixed the Win32 error in the latest commit, but, regardless, I think this PR is ready for feature review.

+@amcastro-tri for feature review please.

---

Review status: 0 of 20 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KPuoV2mhTUF0lgNSeZx:b95f4gf)*
<!-- Sent from Reviewable.io -->

liangfok commented:
This PR passed https://drake-jenkins.csail.mit.edu/job/experimental/4035/ and is thus ready for feature review.

liangfok commented:
This PR passed https://drake-jenkins.csail.mit.edu/job/experimental/4080/ and is thus ready for feature review.

I believe it's blocked on https://github.com/RobotLocomotion/drake/pull/3245.

---

Review status: 0 of 20 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KQ1INTmWNLQD8y-heva:b-nu4ooj)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
@liangfok: I might give you exactly what you need as an output of RBS for input to BotVisualizer. Meaning that you might not even need to instantiate with an RBT at all!
After RBS 2 you'll be able to write an actual test. 
and after @sherm1's simulator you will have a `Publish()` method!
Could you wait another day? :-) 

---

Review status: 0 of 20 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/bot_visualizer_system.cc, line 61 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-KQ2B0pY1NHa032faCa-:-KQ2B0pY1NHa032faCa0:b8wsp2i) ([raw file](https://github.com/robotlocomotion/drake/blob/be1949a3426478374a4a86214722db5283915d13/drake/systems/bot_visualizer_system.cc#L61)):*
> ```cpp
>     auto transform = tree_.relativeTransform(cache, 0, i);
>     auto quat = drake::math::rotmat2quat(transform.linear());
>     std::vector<float>& position = draw_msg_.position[i];
> ```

Great! so what you need as an output of the RBS is a position and a quaternion. I will give you exactly that so you don't need to call `doKinematics` here

---

*[drake/systems/bot_visualizer_system.h, line 103 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-KQ2AH1bzFwk6o6ZasHQ:-KQ2AH1bzFwk6o6ZasHR:bc2t5s9) ([raw file](https://github.com/robotlocomotion/drake/blob/be1949a3426478374a4a86214722db5283915d13/drake/systems/bot_visualizer_system.h#L103)):*
> ```objc
>   // The LCM draw message to send to the Drake Visualizer. This member variable
>   // is declared mutable so it can be modified by EvalOutput().
>   mutable drake::lcmt_viewer_draw draw_msg_;
> ```

This defeats the "glass box" concept of System 2.0. I will make one of the outputs of the RBS a list of the bodies' names for you to use and a vector of poses readily for you to use. There'd be no need for an RBT here! and no fiddling with kinematics cache!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KQ29oz0berl-JetavUz:b-isoyyt)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Of course :wink: 

It would be awesome if you provide `BotVisualizerSystem` with the cache from which the transforms can be derived.

It would also be great if we could write an end-to-end simulation test showing SDF model → `RigidBodyPlantSystem2` → `BotVisualizerSystem2` → `Drake Visualizer, all being run by `Simulation2`.;

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KQ2Bvbs2fTYEWTlaJ0y:b-fx79wj)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/bot_visualizer_system.h, line 103 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-KQ2AH1bzFwk6o6ZasHQ:-KQ2CoDRNoEoTyzH2DVj:b5ywbsm) ([raw file](https://github.com/robotlocomotion/drake/blob/be1949a3426478374a4a86214722db5283915d13/drake/systems/bot_visualizer_system.h#L103)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> This defeats the "glass box" concept of System 2.0. I will make one of the outputs of the RBS a list of the bodies' names for you to use and a vector of poses readily for you to use. There'd be no need for an RBT here! and no fiddling with kinematics cache!
</details>
(1) I think the halfway-initialized `draw_msg_` can be copied locally in `EvalOutput`, so that we don't need mutable here, but ...

(2) BotVisualizerSystem (and LCM in general) is _not_ part of glass box analysis (see how it only uses double, and not any other ScalarType?) -- it is useful for realtime experiments and demo's with real robots, _not_ for analysis.  There _should be no requirement or expectation_ of such code to be glass box.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 20 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/bot_visualizer_system.h, line 103 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-KQ2AH1bzFwk6o6ZasHQ:-KQ2EFOaBneZw-m3JDFo:be6lwls) ([raw file](https://github.com/robotlocomotion/drake/blob/be1949a3426478374a4a86214722db5283915d13/drake/systems/bot_visualizer_system.h#L103)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> (1) I think the halfway-initialized `draw_msg_` can be copied locally in `EvalOutput`, so that we don't need mutable here, but ...
> 
> (2) BotVisualizerSystem (and LCM in general) is _not_ part of glass box analysis (see how it only uses double, and not any other ScalarType?) -- it is useful for realtime experiments and demo's with real robots, _not_ for analysis.  There _should be no requirement or expectation_ of such code to be glass box.
> 
</details>
Is the increased memory utilization  and execution latency caused by instantiating and copying a `drake::lcmt_viewer_draw` message each cycle of the simulation loop a concern? Will the compiler optimize this away?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/bot_visualizer_system.h, line 103 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-KQ2AH1bzFwk6o6ZasHQ:-KQ2FeYg4gjFo99mvpjo:b-j3sonx) ([raw file](https://github.com/robotlocomotion/drake/blob/be1949a3426478374a4a86214722db5283915d13/drake/systems/bot_visualizer_system.h#L103)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Is the increased memory utilization  and execution latency caused by instantiating and copying a `drake::lcmt_viewer_draw` message each cycle of the simulation loop a concern? Will the compiler optimize this away?
</details>
I am unconcerned by at least 6 orders of magnitude.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 20 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/bot_visualizer_system.h, line 103 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-KQ2AH1bzFwk6o6ZasHQ:-KQ2J0nTFMyy3I2WfAyE:bc52tdo) ([raw file](https://github.com/robotlocomotion/drake/blob/be1949a3426478374a4a86214722db5283915d13/drake/systems/bot_visualizer_system.h#L103)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I am unconcerned by at least 6 orders of magnitude.
</details>
:laughing: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
> It would be awesome if you provide BotVisualizerSystem with the cache from which the transforms can be derived

I'll give you the transforms as an output to RBS! isn't that much better? and we'll avoid unnecessary computations!

> It would also be great if we could write an end-to-end simulation test 

Yes! that's the plan for the sprint. I just talked to @naveenoid on hungouts about what components he needed. He will start working on the  System 2 gravity compensator and hopefully with the arrival of #3218 we will be able to do just that! 

---

Review status: 0 of 20 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KQ2IZuG7N565V_FRnYN:ba21fve)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Yes, let me know how I should change this PR to directly accept the transforms.

---

Review status: 0 of 20 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KQ2KHVJC8u42e_gdjRb:ba9tn2r)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Should the system block being introduced in this PR be called `MultiBodyDrakeVisualizerPublisher`? After all, it currently only visualizes rigid bodies. Other visualizers for sensors, actuators, force elements, etc, can be added in the future.

I bring this up because the number of inputs must be declared at the time of construction and currently only the `RigidBodyTree` is passed into the constructor. I think adding sensors, actuators, force elements, etc., into the mix will make things unnecessarily messy. Keeping this system block focused on visualizing just the `RigidBodyTree` is also compatible with #3245, which currently only contains a `RigidBodyTree`.

The name "DrakeVisualizerPublisher" more accurately describes what this system block does -- it does not actually visualize anything; it simply publishes messages that are intended to be consumed by the Drake Visualizer.

---

Review status: 0 of 20 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KQ3pH1-xvFePAC3Msl4:b-rnfvdw)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
I like `DrakeVisualizerPublisher`!

amcastro-tri commented:
I would love to give this a try with `RigidBodyPlant` this weekend (just by forking your branch now for the sprint). @liangfok, could you updated to use `DeclareInputPort`/`DeclareOutputPort` as discussed yesterday in slack?
If you do that then the only `System<T>` method you need to write is `EvalOutput`. The rest have default implementations. 

---

Review status: 0 of 20 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/bot_visualizer_system.h, line 103 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-KQ2AH1bzFwk6o6ZasHQ:-KQ7K2gtHyPe1w8JpJ6x:b-r940e7) ([raw file](https://github.com/robotlocomotion/drake/blob/be1949a3426478374a4a86214722db5283915d13/drake/systems/bot_visualizer_system.h#L103)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> :laughing: 
</details>
Those are very good observations. I would feel a bit more piece of mind if `draw_msg_` was intialized locally as now is done with the `KinematicsCache` in `EvalOutput`, and so not having a `mutable`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KQ7KNpe4Jd2gsob1S_y:b-lfavpd)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Are you OK if a user mistakenly believes that `DrakeVisualizerPublisher` will publish draw commands for things beyond just `RigidBody` transforms? Or, do you expect that this one system block will also eventually publish visualizations of sensors, force elements, etc?

liangfok commented:
Yes, I will update this PR to use `DeclareInputPort()` and `DecareOutputPort()` today. It'll have to be after my ongoing meetings this afternoon, and may have to wait till this evening.

---

Review status: 0 of 20 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KQ7LYtap2EZI4VkeOck:b-e3yjw9)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks.

---

Review status: 0 of 20 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/bot_visualizer_system.cc, line 61 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-KQ2B0pY1NHa032faCa-:-KQ8ea-u7gbTms_QWLSd:b-52yzwb) ([raw file](https://github.com/robotlocomotion/drake/blob/be1949a3426478374a4a86214722db5283915d13/drake/systems/bot_visualizer_system.cc#L61)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Great! so what you need as an output of the RBS is a position and a quaternion. I will give you exactly that so you don't need to call `doKinematics` here
</details>
OK. Perhaps in a future PR since it would basically be an optimization of this one and we're in a sprint?

---

*[drake/systems/bot_visualizer_system.h, line 103 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-KQ2AH1bzFwk6o6ZasHQ:-KQ8fJE3vn8Towpgrr00:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/be1949a3426478374a4a86214722db5283915d13/drake/systems/bot_visualizer_system.h#L103)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Those are very good observations. I would feel a bit more piece of mind if `draw_msg_` was intialized locally as now is done with the `KinematicsCache` in `EvalOutput`, and so not having a `mutable`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KQ8yQ3w8oFzlIqC7xI3:bblnx7v)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 0 of 20 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/systems/bot_visualizer_system.h, line 38 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-KQB1ojsAZtKmc-TgQib:-KQB1ojt_F8eub4nVjXH:b-8od22r) ([raw file](https://github.com/robotlocomotion/drake/blob/cf6143816bb4e36138749c4ea6ebf9ded27874c4/drake/systems/bot_visualizer_system.h#L38)):*
> ```objc
>  */
> class DRAKEBOTVISUALIZERSYSTEM_EXPORT BotVisualizerSystem :
>     public System<double> {
> ```

If you inherit from `LeafSystem` you already have default implementations for CreateDefaultContext, AllocateOutput. So you don't need to write those because you used DeclareInputPort in the constructor and you don't have anything in the context.

---

*[drake/systems/bot_visualizer_system.h, line 68 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-KQB2D_jNMJlb6MT6ZIM:-KQB2D_jNMJlb6MT6ZIN:b-cd2dtl) ([raw file](https://github.com/robotlocomotion/drake/blob/cf6143816bb4e36138749c4ea6ebf9ded27874c4/drake/systems/bot_visualizer_system.h#L68)):*
> ```objc
>    * no state.
>    */
>   std::unique_ptr<ContextBase<double>> CreateDefaultContext() const override;
> ```

Not needed

---

*[drake/systems/bot_visualizer_system.h, line 73 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-KQB2EpA8r_dWK9OOl98:-KQB2EpA8r_dWK9OOl99:b-cd2dtl) ([raw file](https://github.com/robotlocomotion/drake/blob/cf6143816bb4e36138749c4ea6ebf9ded27874c4/drake/systems/bot_visualizer_system.h#L73)):*
> ```objc
>    * The output contains zero ports.
>    */
>   std::unique_ptr<SystemOutput<double>> AllocateOutput(
> ```

Not needed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks!

---

Review status: 0 of 20 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/bot_visualizer_system.h, line 38 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-KQB1ojsAZtKmc-TgQib:-KQCFakFPImbhoyUU0YY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cf6143816bb4e36138749c4ea6ebf9ded27874c4/drake/systems/bot_visualizer_system.h#L38)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> If you inherit from `LeafSystem` you already have default implementations for CreateDefaultContext, AllocateOutput. So you don't need to write those because you used DeclareInputPort in the constructor and you don't have anything in the context.
</details>
Done.

---

*[drake/systems/bot_visualizer_system.h, line 68 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-KQB2D_jNMJlb6MT6ZIM:-KQCG3hzIg8DckjLcavd:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cf6143816bb4e36138749c4ea6ebf9ded27874c4/drake/systems/bot_visualizer_system.h#L68)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Not needed
</details>
Done.

---

*[drake/systems/bot_visualizer_system.h, line 73 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-KQB2EpA8r_dWK9OOl98:-KQCG4OE06qaTggylp6w:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cf6143816bb4e36138749c4ea6ebf9ded27874c4/drake/systems/bot_visualizer_system.h#L73)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Not needed
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KQCHzKFLFkKl40895ru:bblnx7i)*
<!-- Sent from Reviewable.io -->

liangfok commented:
The unit tests in this PR can be further polished by extracting common code into a single method. I shall do that prior to merging this PR and shall thus apply the "do not merge" label.

liangfok commented:
I finished polishing the unit tests and will thus remove the "do not merge label".

sherm1 commented:
Should use Publish() rather than EvalOutput().

---

Review status: 0 of 22 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/bot_visualizer_system.h, line 70 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-KQSG58kQNXOsgiTvbZd:-KQSG58l2maoYJKfCy23:b-l0oxie) ([raw file](https://github.com/robotlocomotion/drake/blob/dc399900986f901449528787cc4999c41ad768ba/drake/systems/bot_visualizer_system.h#L70)):*
> ```objc
>    * the current state of the model in Drake Visualizer.
>    */
>   void EvalOutput(const ContextBase<double>& context,
> ```

This System should implement Publish() but not EvalOutput() (since it has no output ports).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KQSGFNCF1FZTvmr21UX:b-odls4w)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks!

---

Review status: 0 of 22 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/bot_visualizer_system.h, line 70 \[r11\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-KQSG58kQNXOsgiTvbZd:-KQT22mQbseUqU2Wp83h:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/dc399900986f901449528787cc4999c41ad768ba/drake/systems/bot_visualizer_system.h#L70)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> This System should implement Publish() but not EvalOutput() (since it has no output ports).
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KQT23s_gFxmRl63pd1d:bblnx7i)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
I gave this a quick skim.  I think the quantity of code and test infrastructure is overly much for what it is accomplishing.  I think the System block, as well its the unit testing, could be more abbreviated.

The draw message method is really just an `LcmAndVectorBaseTranslator` from the RBT output vector to the draw message -- in which case, this is just is-a or has-a `LcmPublisherSystem` and all of the `LCM` mechanics are reused instead of reimplemented.

The load robot message could be similarly factored out as a "give me the bytes", and so tested without actually sending LCM.  Its drake-shape dispatch could be further exposed for direct unit testing, instead of having to round-trip a full load_robot message.  I suspect a visitor or archive pattern added to the shapes would allow for type-based case dispatch and less code as well.

The unit tests should not be loading sdfs (the parser should be out of scope); the RBTs should be constructed programmatically.  It would be nice in fact, if this class _just_ took in the position information as a vector, and published the draw message -- without having to consult the kinematics cache itself.  Even today, there could be a separate System that output the link position information, and this class merely formats it into a draw message.

At most, one christmass-tree acceptance test that loads a file and simulates could serve as a regression guard (for this code, as well as likely the System 2.0, Simulator 2.0, etc.).

liangfok commented:
Interesting. Another strategy @amcastro-tri and I were planning to take, which would also greatly reduce the code in this PR, is to have `RigidBodySystem 2.0` output the transforms for every `RigidBodyTree` in the world, which `BotVisualizerSystem` would then directly publish (no need to even maintain a `RigidBodyTree` internally. Which route do you think is better?

jwnimmer-tri commented:
That's an improvement, but essentially just the _"Even today, there could be a separate System that output the link position information, and this class merely formats it into a draw message."_ portion of my post.  (The load_robot portion would still need the tree, though, perhaps?)  The verbose and non-unit nature of the test programs should still be addressed (IMO).

liangfok commented:
Yes, a reference to `RigidBodyTree` will be needed to create the `lcmt_viewer_load_robot` message. This is done once during the time of construction (see `BotVisualizerSystem::initialize_drake_visualizer()`). During the run-time, if `RigidBodySystem2.0` provides `BotVisualizer2.0` with transform data, a reference to `RigidBodyTree` will not be needed.

amcastro-tri commented:
As far as I understand, since I still didn't have the time to do a full review of this PR, you only need rigid bodies names and poses. My plan is to create a Sys 2.0 block (maybe called `RbpStateToBodyPoses`) that takes a state vector (it could come from either RBS 2.0 or from a simple pendulum like example not using RBS 2.0 but using BotVisualizer for visualization of its state). The `RbpStateToBodyPoses` would also output rigid body names.

So I think `BotVisualizer` could have as inputs a channel for rigid body names and a channel for poses. With that information it puts together LCM messages and send them through. No need for `doKinematics` or any other RBT computation to be duplicated.

Does this make any sense?

amcastro-tri commented:
I just created #3316 which will take the input of an RBP and output names and poses. This information could then be used by `BotVisualizer` or any other system needing this information.

liangfok commented:
> It would be nice in fact, if this class just took in the position information as a vector, and published the draw message -- without having to consult the kinematics cache itself.  Even today, there could be a separate System that output the link position information, and this class merely formats it into a draw message.

Yes. I will apply the "do not review label to this PR until `RigidBodyPlant` is in master and outputs said rigid body pose information.

 > The draw message method is really just an `LcmAndVectorBaseTranslator` from the RBT output vector to the draw message -- in which case, this is just is-a or has-a `LcmPublisherSystem` and all of the LCM mechanics are reused instead of reimplemented.

OK I will see if this change makes sense after `RigidBodyPlant` outputs the body pose transforms.

> The unit tests should not be loading sdfs (the parser should be out of scope); the RBTs should be constructed programmatically.

Once `RigidBodyPlant` outputs body pose information, there will not be a need to have a `RigidBodyTree` in `BotVisualizerSystem 2.0` and thus will eliminate the need to instantiate one whether programmatically or via SDF.

---

Review status: 0 of 22 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KQ_NMLhrf081JunpOQd:b-rp8xmm)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Based on https://github.com/RobotLocomotion/drake/issues/3325#issuecomment-246110353, I'm now thinking the System 2.0 version of `BotVisualizer` should take as input poses that are outputted by sensors in the System 2.0 version of `RigidBodySystem`. I will wait for the System 2.0 version of `RigidBodySystem` to exist in master before updating this PR.

liangfok commented:
`BotVizualizer 2.0` should be updated to wait for a `DRAKE_VIEWER_STATUS` message prior to sending a `DRAKE_VIEW_LOAD_ROBOT` message. See: https://github.com/RobotLocomotion/drake/issues/3421#issuecomment-246452325.

rpoyner-tri commented:
@liangfok I am not sure the wait for the viewer status is a good idea inside BotVisualizer 2.0; the message you are waiting for also only gets sent once, so you would have just inverted the hazard of #3421. That is why I recommended a wait-for-message in the *launcher*, not in the simulation app.

liangfok commented:
@rpoyner-tri: I see. I did not realize the `DRAKE_VIEWER_STATUS` message is only sent once. Maybe it should be updated to be periodically sent to detect and recover from random UDP message loss?

liangfok commented:
Now that #3245 is merged, I shall resume work on this first thing in the morning.

---

Review status: 0 of 22 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KQWs8eOuiTEfw6ae026:b-nlpf45)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot retest this please

liangfok commented:
@jwnimmer-tri: Here's my new plan, what do you think?

I'm going to make `BotVisualizer` (1) maintain a `LcmPublisherSystem` as a private member variable, (2) creates for itself an object of type `TranslatorBetweenLcmtViewerDraw` that is uses with the `LcmPublisherSystem`, and (3) has custom code that publishes a `drake::lcmt_viewer_load_robot` once at startup in a manner similar to `drake::cars::BotVisualizerHack`'s  `PublishLoadRobot()`.

Do you agree?

---

Review status: 0 of 22 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/systems/bot_visualizer_system.cc, line 61 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-KQ2B0pY1NHa032faCa-:-KRd9rqPWEiu-b76_Sky:b-f3ip6a) ([raw file](https://github.com/robotlocomotion/drake/blob/be1949a3426478374a4a86214722db5283915d13/drake/systems/bot_visualizer_system.cc#L61)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> OK. Perhaps in a future PR since it would basically be an optimization of this one and we're in a sprint?
</details>
@amcastro-tri: Since #3245 was merged without outputting the pose information, can we defer the simplification of this system till a future PR?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KRdECFPvy5doZCl0m58:b-3kvg5z)*
<!-- Sent from Reviewable.io -->

liangfok commented:
An alternative to having custom code for publishes the `drake::lcmt_viewer_load_robot` message is to instantiate a _second_ `LcmPublisherSystem` that uses a `TranslatorBetweenLcmtViewerLoadRobot` to publish the `drake::lcmt_viewer_load_robot` message, but that may be taking things a bit too far. 

Thoughts?

---

Review status: 0 of 22 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KRdH3Ch0EGc8KI2hYIv:b19xfil)*
<!-- Sent from Reviewable.io -->

liangfok commented:
_(Sorry for the huge number of spelling / grammar errors; thinking & typing quickly here...)_

---

Review status: 0 of 22 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KRdHkCVOYdpTuGJmZXn:bmhuj7j)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
I will work today on having output ports that you can use in the `BotVisualizer` (RigidBodyTreeLcmPublisher?).

liangfok commented:
@amcastro-tri: Thanks! OK I will use the name `RigidBodyTreeLcmPublisher` instead of `BotVisualizer`.

From F2F discussion with @jwnimmer-tri, my updated plan is to make `RigidBodyTreeLcmPublisher` a subclass of `LcmPublisherSystem` that (1) uses a `TranslatorBetweenLcmtViewerDraw` as the translator and (2) contains bespoke code to publish the `drake::lcmt_viewer_load_robot` message.

I shall now close this PR and issue new PRs based on the new design.

---

Review status: 0 of 22 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3228#-:-KRdejh01aqcxamijG9G:b-2rozqx)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
I trust you and @jwnimmer-tri on this design since you are the LCM experts. So what are you expecting the inputs to `RigidBodyTreeLcmPublisher` to be? is my plan of having an output in RBP for rigid body kinematics information sound to you?

liangfok commented:
Yes! If you can modify `RigidBodyPlant` to output pose information for each body, I will make `RigidBodyTreeLcmPublisher` consume it.

If `RigidBodyPlant` can also provide visualization information for each body, then `RigidBodyTreeLcmPublisher` will not even need to have a `RigidbodyTree`.

amcastro-tri commented:
> If RigidBodyPlant can also provide visualization information for each body, then RigidBodyTreeLcmPublisher will not even need to have a RigidbodyTree.

I know! that'd be ideal. However I think as a good compromise to get this going is to initialize now `RigidBodyTreeLcmPublisher` from an RBT and then make it consume RBP's output. I think that is a good compromise until we decide how exactly to output this geometry information.

liangfok commented:
Agreed. Baby steps....

Here's to incremental PRs!

RussTedrake commented on drake/systems/bot_visualizer_system.h:
```diff
@@ -0,0 +1,103 @@
+ * To support multiple simultaneous instances of this system, which may
+ * occur when multiple unit tests are running in parallel, the channel names
+ * can be modified by a postfix string, which is specified as an input parameter
+ * to the constructor.
```
please add @ingroup systems to make it show up here:
http://drake.mit.edu/doxygen_cxx/group__systems.html
example here: https://github.com/RobotLocomotion/drake/blob/master/drake/systems/framework/primitives/adder.h#L16
@amcastro-tri -- would be good to do this for RigidBodyPlant, too

liangfok replied:
Done. See: https://github.com/RobotLocomotion/drake/pull/3494/files#diff-389775f8f832bbac71309d353a2b22e5R39
